modules/ctl/src/main/scala/free/ctl.scala

```diff
@@ -0,0 +1,124 @@
+        case Server.Remote(Host.Network(h), u)  => u.foldRight(h)(_ + "@" + _)
+      }
+  }
+  object Server {
+    case object Local extends Server
+    case class  Remote(host: Host, user: Option[String]) extends Server
+  }
+
+
+  /** ADT for output log levels. */
+  sealed trait Level
+  object Level {
+    case object Info  extends Level
+    case object Warn  extends Level
+    case object Error extends Level
+    case object Shell extends Level
+  }
+
+  /** ADT of low-level control operations. */
+  sealed trait CtlOp[A]
+  object CtlOp {
+    case class  Exit[A](exitCode: Int)                              extends CtlOp[A]
+    case object GetConfig                                           extends CtlOp[Config]
+    case class  Gosub[A](level: Level, msg: String, fa: CtlIO[A])   extends CtlOp[A]
+    case class  Shell(remote: Boolean, cmd: String \/ List[String]) extends CtlOp[Output]
+  }
+
+  /** Combinator that applies a check to command output. */
+  def require[A](shell: CtlIO[Output])(f: PartialFunction[Output, A]): CtlIO[A] =
+    shell.flatMap { o =>
+      f.lift(o) match {
+        case None    =>
+          o.lines.traverse(error(_))      *>
+          error(s"exited (${o.exitCode})") *>
+          exit[A](o.exitCode)
+        case Some(a) => a.point[CtlIO]
+      }
+    }
+
+  /** Syntax to allow checks on shell command output. */
+  implicit class CtlIOOps[A](fa: CtlIO[A]) {
+    def require[B](f: PartialFunction[Output, B])(implicit ev: A =:= Output): CtlIO[B] =
+      ctl.require(fa.map(ev))(f)
+  }
+
+  /** Free monad over CtlOp. */
+  type CtlIO[A] = Free[CtlOp, A]
+
+  // Constructors
+
+  def serverHostName: CtlIO[String] =
+    server.flatMap {
+      case Server.Local => "localhost".point[CtlIO]
+      case Server.Remote(Host.Network(name), _) => name.point[CtlIO]
+      case Server.Remote(Host.Machine(name), _) =>
+        shell("docker-machine", "ip", name).require {
+          case Output(0, s :: Nil) => s
+        }
+    }
+
+  def shell(c: String): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(false, c.left))
+
+  def shell(c: String, cs: String*): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(false, (c :: cs.toList).right))
+
+  def remote(c: String): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(true, c.left))
+
+  def remote(c: String, cs: String*): CtlIO[Output] =
+    Free.liftF(CtlOp.Shell(true, (c :: cs.toList).right))
+
+  def exit[A](exitCode: Int): CtlIO[A] =
+    Free.liftF(CtlOp.Exit[A](exitCode))
+
+  val config: CtlIO[Config] =
+    Free.liftF(CtlOp.GetConfig)
+
+  val server: CtlIO[Server] =
+    config.map(_.server)
+
+  val isRemote: CtlIO[Boolean] =
+    server.map {
+      case Server.Remote(_, _) => true
+      case Server.Local        => false
+    }
+
+  def gosub[A](msg: String)(fa: CtlIO[A]): CtlIO[A] =
+    Free.liftF(CtlOp.Gosub(Level.Info, msg, fa))
+
+  def logMessage(level: Level, msg: String): CtlIO[Unit] =
+    Free.liftF(CtlOp.Gosub(level, msg, ().point[CtlIO]))
```

Comment:
It's constructing a program that will log the message when it's interpreted, and logging a simple message is a special case of logging a message and then running a sub-program at a higher indentation so there's some extra cruft in there. But it keeps the algebra simple because there's only one log constructor.